### PR TITLE
implement instant search and its e2e test

### PR DESCRIPTION
**Pull request by aceTmobile**
- user can now see search results as they type
- throttle call to API by 500ms

### CODE REVIEW

File: libs/books/feature/src/lib/book-search/book-search.component.html
Comment (catfireparty) [inline libs/books/feature/src/lib/book-search/book-search.component.html]: While this kind of works, there is a better way to get the values from the control.

File: libs/books/feature/src/lib/book-search/book-search.component.ts
Comment (catfireparty) [inline libs/books/feature/src/lib/book-search/book-search.component.ts]: Have you looked at the console while this runs? If you type `Javascript` into the box it makes a call for every letter and throws proxy errors. This almost certainly isn't what you want.

Look into `FormControl.valueChanges` and check out `debounceTime` and maybe even `distinctUntilChanged`.

File: libs/books/feature/src/lib/book-search/book-search.component.ts
Comment (catfireparty) [inline libs/books/feature/src/lib/book-search/book-search.component.ts]: What happens when this control is destroyed? Isn't this a memory leak?

File: libs/books/feature/src/lib/book-search/book-search.component.html
Comment (aceTmobile) [reply libs/books/feature/src/lib/book-search/book-search.component.html]: updated

File: libs/books/feature/src/lib/book-search/book-search.component.ts
Comment (aceTmobile) [reply libs/books/feature/src/lib/book-search/book-search.component.ts]: updated

File: libs/books/feature/src/lib/book-search/book-search.component.ts
Comment (aceTmobile) [reply libs/books/feature/src/lib/book-search/book-search.component.ts]: updated

File: libs/books/feature/src/lib/book-search/book-search.component.ts
Comment (mandarini) [inline libs/books/feature/src/lib/book-search/book-search.component.ts:37]: Why did you choose to use this lifecycle method? Is it really needed? Why yes, why no?

**Review by mandarini — CHANGES_REQUESTED** [review]
Please fix merge conflicts (the branch needs to be mergable), squash commits :)

File: libs/books/feature/src/lib/book-search/book-search.component.ts
Comment (aceTmobile) [reply libs/books/feature/src/lib/book-search/book-search.component.ts:37]: Why did you choose to use this lifecycle method? Is it really needed?
my answer: I would say that it depends on whether or not this code is deemed an initialization task or not. Since I was in between I chose afterViewInit simply because this code listens for user changes that happen after the view is initialized and felt this LCE was more appropriate. An argument could be made that AfterViewInit is ideal for handling tasks that requires/ensures that its child components are initialized before executing which isn't the case here. I don't have any strong thoughts here as I am not well versed in angular and would defer to you & those with more experience in angular.  if you would like to expand on the angular documentation I'm all ears.

**Comment by aceTmobile** [comment]
@mandarini Hi, all conflicts resolved. Please re-review. thanks. 
I typically use the squash and merge button to squash commits before merging. Any issues with that?

File: libs/books/feature/src/lib/book-search/book-search.component.ts
Comment (mandarini) [reply libs/books/feature/src/lib/book-search/book-search.component.ts:37]: Thank you for the explanation. Indeed, the afterViewInit is usually used when you want to make sure that all view childs are there. In any case, we can leave this there now, but I would suggest putting it in ngOnInit next time!

**Comment by mandarini** [comment]
So, it's better to squash the commits on your end, and make sure all looks good and there are no conflicts while squashing, and submit an MR with only one commit. This gives you better control and understanding of how the actual squashing is going to happen, also you get better control over the commit message, etc. I would suggest you squash, since this is how the MRs are also submitted in the actual Tmo repository, so it would be nice practice. Let me know if you need help.

**Comment by aceTmobile** [comment]
@mandarini Hi I was able to squash commits but ran into conflicts after adding spacing. Does this work? If no, then yes I definitely need a little help on this thanks.

**Comment by mandarini** [comment]
Hi @aceTmobile , let me know how I can help. Yes, squashing commits can result in conflicts, that's why it's best to do it locally, so that you know which code you want to keep.

**Comment by mandarini** [comment]
Also, it looks like you merged `chore/code-review` into your branch, instead of rebasing. Let me know if you need help with rebasing, too!

**Comment by aceTmobile** [comment]
@mandarini hi, I made corrections. The mr should be good now. Please re-review. Thanks.

**Comment by mandarini** [comment]
Looks good to me thanks

**Pull request merged**
